services/auth/source/ldap/source_search.go

```diff
@@ -192,6 +196,71 @@ func checkRestricted(l *ldap.Conn, ls *Source, userDN string) bool {
 	return false
 }
 
+// List all group memberships of a user
+func (ls *Source) listLdapGroupMemberships(l *ldap.Conn, uid string) []string {
+	var ldapGroups []string
+	groupFilter := fmt.Sprintf("(%s=%s)", ls.GroupMemberUID, uid)
```

Comment:
It would be nice if the filter string was configurable.

For example, our organization doesn't store usernames in the group entries and instead stores the full distinguished name (dn) of the user. So with the current code, our search string would be `"(%s=cn=%s,ou=people,dc=example,dc=com)"`.
With the possible use of [explicit argument indexes](https://pkg.go.dev/fmt#hdr-Explicit_argument_indexes) in fmt, all cases should be covered after this change.
If you leave the current format string as the default value, there should be no difference for most use cases.

I would also like to mention that with the current code, the ldap user has to have the permission to see in which groups he is (this may not be the case in every configuration). Especially with the use of a bind dn, one would expect that it is also used for group search and filtering. This may be a bug.

Otherwise, this change works for us. Thank you for your work on it!

Reply:
`uid` here is the user's DN, not username...

IMO variables for this package should be renamed... They seem a bit misleading. Like, `uid` is a valid attribute in some ldap implementations, but here it is `DN`, which is another attribute.
Another good example is `GroupMemberUID` which is in fact the name of the User's attribute that is stored in group... A better name in my view would be `UserAttrInGroup` to express that is is a user attribute stored in a group...

Reply:
> It would be nice if the filter string was configurable.

I'd say that if we make the filter configurable, we can drop the GroupMemberUID config option, because this could then be hard coded into the filter.

> I would also like to mention that with the current code, the ldap user has to have the permission to see in which groups he is (this may not be the case in every configuration). Especially with the use of a bind dn, one would expect that it is also used for group search and filtering. This may be a bug.

Regarding the Bind user: I'll look into it. AFAICT the bind user should already be used, because we do not have access to the user password in most cases and therefore cannot use it for authentication.

Reply:
> `uid` here is the user's DN, not username...

It can actually be both. But we can change the variable name to reflect this.